zerver/models.py

```diff
@@ -894,17 +899,21 @@ def web_public_streams_enabled(self) -> bool:
         return True
 
     def has_web_public_streams(self) -> bool:
-        """
-        If any of the streams in the realm is web
-        public, then the realm is web public.
-        """
         if not self.web_public_streams_enabled():
             return False
 
         from zerver.lib.streams import get_web_public_streams_queryset
 
         return get_web_public_streams_queryset(self).exists()
 
+    def allow_web_public_streams_access(self) -> bool:
+        """
+        If any of the streams in the realm is web
+        public and `enable_spectator_access` is True,
+        then the Realm is web public.
+        """
+        return self.enable_spectator_access and self.has_web_public_streams()
```

Comment:
Should the UserProfile logic in `can_create_web_public_streams` be checking this rather than `web_public_streams_enabled`?  I think the answer is yes.

Reply:
But, we check `get_web_public_streams_queryset(self).exists()` as part of `allow_web_public_streams_access`; This means a user cannot create a web public stream unless there is already one created.

Reply:
I think the right model is that you should first need to enable `enable_spectator_access`, and then it should be possible to create a web-public stream, and only then should the spectator login button appear.